pulsar-io/elastic-search/src/test/java/org/apache/pulsar/io/elasticsearch/ElasticSearchExtractTests.java

```diff
@@ -286,4 +287,144 @@ public Object getNativeObject() {
+            keySchemaBuilder.field("b").type(SchemaType.STRING).optional().defaultValue(null);
+            keySchemaBuilder.field("c").type(SchemaType.STRING).optional().defaultValue(null);
+
+            RecordSchemaBuilder udtSchemaBuilder = SchemaBuilder.record("type1");
+            udtSchemaBuilder.field("b_inside_inner").type(SchemaType.STRING).optional().defaultValue(null);
+            udtSchemaBuilder.field("a_inside_inner").type(SchemaType.STRING).optional().defaultValue(null);
+            GenericSchema<GenericRecord> udtGenericSchema = GenericSchemaImpl.of(udtSchemaBuilder.build(schemaType));
+
+            keySchemaBuilder.field("inner", udtGenericSchema).type(schemaType).optional().defaultValue(null);
+            GenericSchema<GenericRecord> keySchema = GenericSchemaImpl.of(keySchemaBuilder.build(schemaType));
+
+            final GenericRecord innerRecord = udtGenericSchema.newRecordBuilder()
+                    .set("b_inside_inner", "0b_value_from_inner")
+                    .set("a_inside_inner", "a_value_from_inner")
+                    .build();
+
+            GenericRecord keyGenericRecord = keySchema.newRecordBuilder()
+                    .set("c", "c_key")
+                    .set("b", "0b_key")
+                    .set("a", "a_key")
+                    .set("inner", innerRecord)
+                    .build();
+
+            Record<GenericObject> genericObjectRecord = getKeyValueGenericObject(schemaType, keySchema, keyGenericRecord);
+
+
+            ElasticSearchSink elasticSearchSink = new ElasticSearchSink();
+            elasticSearchSink.open(ImmutableMap.of(
+                    "elasticSearchUrl", "http://localhost:9200",
+                    "compatibilityMode", "ELASTICSEARCH",
+                    "schemaEnable", "true",
+                    "canonicalKeyFields", "false",
+                    "keyIgnore", "false"), null);
+            Pair<String, String> pair = elasticSearchSink.extractIdAndDocument(genericObjectRecord);
+            assertEquals(pair.getKey(), "[\"a_key\",\"0b_key\",\"c_key\",{\"b_inside_inner\":\"0b_value_from_inner\",\"a_inside_inner\":\"a_value_from_inner\"}]");
+
+            elasticSearchSink = new ElasticSearchSink();
+            elasticSearchSink.open(ImmutableMap.of(
+                    "elasticSearchUrl", "http://localhost:9200",
+                    "compatibilityMode", "ELASTICSEARCH",
+                    "schemaEnable", "true",
+                    "canonicalKeyFields", "true",
+                    "keyIgnore", "false"), null);
+            pair = elasticSearchSink.extractIdAndDocument(genericObjectRecord);
+            assertEquals(pair.getKey(), "[\"a_key\",\"0b_key\",\"c_key\",{\"a_inside_inner\":\"a_value_from_inner\",\"b_inside_inner\":\"0b_value_from_inner\"}]");
+        }
+
+        {
+            RecordSchemaBuilder keySchemaBuilder = org.apache.pulsar.client.api.schema.SchemaBuilder.record("key");
```

Comment:
Can this go to a distinct test with explicit naming explaining the difference of what we are testing.
Also not a big fan of scoping inside functions.